### PR TITLE
Add assert.h to threads_pthread.c for NonStop thread compiles.

### LINKED ISSUE #15809: Missing assert on NonStop x86/ia64 PUT build at beta1
The Posix User Thread build for NonStop is missing an include to support `assert()`
```
c99: running /usr/cmplr/ccomfe -Z8 -ZBL --c99 -ZO2 -Zg3 -ZS/tmp/t503316511ABBFGB.T -D_XOPEN_SOURCE -ZR8 -m -Zf -ZT -D_TANDEM_SOURCE -ZNW203 -ZNW220 -ZNW272 -ZNW734 -ZNW770 -ZNW1506 -Zz -I. -Iinclude -Iproviders/common/include -Iproviders/implementations/include -I/usr/include -DOPENSSL_PIC -DOPENSSLDIR="/usr/local-ssl3.0_put/ssl" -DENGINESDIR="/usr/local-ssl3.0_put/lib/engines-3" -DMODULESDIR="/usr/local-ssl3.0_put/lib/ossl-modules" -DOPENSSL_BUILDING_OPENSSL -DOPENSSL_VPROC= -D_XOPEN_SOURCE -D_XOPEN_SOURCE_EXTENDED=1 -D_TANDEM_SOURCE -DB_ENDIAN -D_TANDEM_ARCH=3 -D_PUT_MODEL_ -D_REENTRANT -D_THREAD_SUPPORT_FUNCTIONS -DNDEBUG -D_REENTRANT crypto/threads_pthread.c
          assert(errno != EDEADLK && errno != EBUSY);
          ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/crypto/threads_pthread.c", line 83: error(114):
          identifier "assert" is undefined

          assert(errno != EDEADLK && errno != EBUSY);
          ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/crypto/threads_pthread.c", line 98: error(114):
          identifier "assert" is undefined

          assert(errno != EPERM);
          ^
"/home/ituglib/randall/jenkins/.jenkins/workspace/OpenSSL-3.0_Pipeline/crypto/threads_pthread.c", line 113: error(114):
          identifier "assert" is undefined
```
The configuration is:
```
Command line (with current working directory = .):

    perl ./Configure nonstop-nsx_64_put --prefix=/usr/local-ssl3.0_put --openssldir=/usr/local-ssl3.0_put/ssl --with-rand-seed=rdcpu threads enable-fips -D_REENTRANT

Perl information:

    perl
    5.26.3 for NSV-D-nonstop_kernel

Enabled features:

    aria
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    egd
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    idea
    legacy
    loadereng
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    threads
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    acvp-tests          [default]        OPENSSL_NO_ACVP_TESTS
    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    asm                 [no asm_arch]    OPENSSL_NO_ASM
    buildtest-c++       [default]
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]
    zlib-dynamic        [default]
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "c99",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    bn_ops => "SIXTY_FOUR_BIT",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-g -O2 -Wextensions -Wnowarn=203,220,272,734,770,1506 -Wbuild_neutral_library -Wverbose -Wtarget=tns/x -Wsystype=oss -Wlp64 -WIEEE_float",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "_TANDEM_SOURCE", "B_ENDIAN", "_TANDEM_ARCH=3", "_PUT_MODEL_", "_REENTRANT", "_THREAD_SUPPORT_FUNCTIONS" ],
    disable => [  ],
    dso_scheme => "DLFCN",
    enable => [ "egd" ],
    ex_libs => "-lrld -lput",
    includes => [  ],
    lflags => "-Wxld=\"-set systype oss\" -Wxld=\"-set data_model lp64\" -Wxld=\"-set floattype ieee_float\"",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "-Wshared",
    perl => "/usr/bin/perl",
    perl_platform => "Unix",
    shared_argfileflag => "-Wxld_obey=",
    shared_cflag => "",
    shared_defflag => "-Wxld_obey=",
    shared_defines => [  ],
    shared_extension => ".so",
    shared_ldflag => "-Wshared",
    shared_rcflag => "",
    shared_target => "nonstop-shared",
    sys_id => "TANDEM",
    thread_defines => [  ],
    thread_scheme => "(unknown)",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    BUILDFILE =
    CC =
    CFLAGS =
    CPPFLAGS =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LDFLAGS =
    LDLIBS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    CC              = c99
    CFLAGS          =
    CPPDEFINES      = _REENTRANT
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

### PR DESCRIPTION
**Pull request by rsbeckerca**
Fixes #15809

Signed-off-by: Randall S. Becker <rsbecker@nexbridge.com>

### CODE REVIEW

**Comment by rsbeckerca** [comment]
This PR also applies to Standard Posix Thread (SPT) builds on NonStop.

File: crypto/threads_pthread.c [hunk 1]
```diff
@@ -24,6 +24,10 @@
 #  include <unistd.h>
 #endif
 
+# if defined(OPENSSL_SYS_TANDEM)
+#  include <assert.h>
+# endif
```
Comment (t8m) [inline crypto/threads_pthread.c:29]: IMO this should be included not just on OPENSSL_SYS_TANDEM but everywhere. Actually it is not needed if USE_RWLOCK is defined below, but IMO it is harmless to always include it.

Comment (rsbeckerca) [reply crypto/threads_pthread.c:29]: That's what I thought, but apparently the CI builds don't seem to need it. Unless the implicit definition is not noticed. So remove the #if/#endif?

Comment (levitte) [reply crypto/threads_pthread.c:29]: Agreed

Comment (t8m) [reply crypto/threads_pthread.c:29]: >  apparently the CI builds don't seem to need it

Most probably because the USE_RWLOCK is being defined for them.

**Comment by rsbeckerca** [comment]
The single line PR is now ready ;)

**Review by levitte — APPROVED** [review]

File: crypto/threads_pthread.c [hunk 2]
```diff
@@ -24,6 +24,8 @@
 #  include <unistd.h>
 #endif
 
+# include <assert.h>
+
 # ifdef PTHREAD_RWLOCK_INITIALIZER
 #  define USE_RWLOCK
 # endif
```
Comment (paulidale) [inline crypto/threads_pthread.c:31]: Beat me to this one :)

```c
# else
#  include <assert.h>
```

is the other way of doing this.  We're only using assert() on the non _USE_RWLOCK_ paths.

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by rsbeckerca** [comment]
May I request that this get merged soon? Our CI pipeline for threads is blocked without it and I would like to test with the current set of merges.

**Comment by t8m** [comment]
This was approved many days ago we just forgot to set the label. Merging.

**Comment by t8m** [comment]
Merged. Thank you for the contribution and patience.

**Pull request closed without merging**